File: benchmarks/Ruby/deltablue.rb
# frozen_string_literal: true

# The benchmark in its current state is a derivation from the SOM version,
# which is derived from Mario Wolczko's Smalltalk version of DeltaBlue.
#
# The original license details are availble here:
# http://web.archive.org/web/20050825101121/http://www.sunlabs.com/people/mario/java_benchmarking/index.html

# This file itself, and its souce control history is however based on the
# following. It is unclear whether this still bears any relevance since the
# nature of the code was essentially reverted back to the Smalltalk version.
#
# Port of deltablue.py, as documented below, to Ruby.
# Stefan Marr, 2014-04-28
#
# Was: deltablue.py
# =================
#
# Ported for the PyPy project.
# Contributed by Daniel Lindsley
#
# This implementation of the DeltaBlue benchmark was directly ported
# from the `V8's source code`_, which was in turn derived
# from the Smalltalk implementation by John Maloney and Mario
# Wolczko. The original Javascript implementation was licensed under the GPL.
#
# It's been updated in places to be more idiomatic to Python (for loops over
# collections, a couple magic methods, OrderedCollection being a list & things
# altering those collections changed to the builtin methods) but largely retains
# the layout & logic from the original. (Ugh.)
#
# .. _`V8's source code`: (http://code.google.com/p/v8/source/browse/branches/bleeding_edge/benchmarks/deltablue.js)

require_relative 'benchmark'
require_relative 'som'

class DeltaBlue < Benchmark
  def inner_benchmark_loop(inner_iterations)
    Planner.chain_test(inner_iterations)
    Planner.projection_test(inner_iterations)
    true
  end
end

class Plan < Vector
  def initialize
    super(15)
  end

  def execute
    each { |c| c.execute }
  end
end

class Planner
  def initialize
    @current_mark = 1
  end

  def incremental_add(constraint)
    mark = new_mark
    overridden = constraint.satisfy(mark, self)

    overridden = overridden.satisfy(mark, self) while overridden
  end

  def incremental_remove(constraint)
    out_v = constraint.output
    constraint.mark_unsatisfied
    constraint.remove_from_graph
    unsatisfied = remove_propagate_from(out_v)
    unsatisfied.each { |u| incremental_add(u) }
  end

  def extract_plan_from_constraints(constraints)
    sources = Vector.new

    constraints.each do |c|
      sources.append(c) if c.is_input && c.is_satisfied
    end

    make_plan(sources)
  end

  def make_plan(sources)
    mark = new_mark
    plan = Plan.new
    todo = sources

    until todo.empty?
      c = todo.remove_first

      if c.output.mark != mark && c.inputs_known(mark)
        plan.append(c)
        c.output.mark = mark
        add_constraints_consuming_to(c.output, todo)
      end
    end

    plan
  end

  def propagate_from(v)
    todo = Vector.new
    add_constraints_consuming_to(v, todo)

    until todo.empty?
      c = todo.remove_first
      c.execute
      add_constraints_consuming_to(c.output, todo)
    end
  end

  def add_constraints_consuming_to(v, coll)
    determining_c = v.determined_by

    v.constraints.each do |c|
      coll.append(c) if (!c.equal? determining_c) && c.is_satisfied
    end
  end

  def add_propagate(c, mark)
    todo = Vector.with(c)

    until todo.empty?
      d = todo.remove_first
      if d.output.mark == mark
        incremental_remove(c)
        return false
      end
      d.recalculate
      add_constraints_consuming_to(d.output, todo)
    end
    true
  end

  def change_var(var, val)
    edit_constraint = EditConstraint.new(var, SYM_PREFERRED, self)
    plan = extract_plan_from_constraints(Vector.with(edit_constraint))
    10.times do
      var.value = val
      plan.execute
    end
    edit_constraint.destroy_constraint(self)
  end

  def constraints_consuming(v)
    determining_c = v.determined_by
    v.constraints.each do |c|
      yield c if c != determining_c && c.is_satisfied
    end
  end

  def new_mark
    @current_mark += 1
  end

  def remove_propagate_from(out_v)
    unsatisfied = Vector.new

    out_v.determined_by = nil
    out_v.walk_strength = ABSOLUTE_WEAKEST
    out_v.stay = true

    todo = Vector.with(out_v)

    until todo.empty?
      v = todo.remove_first

      v.constraints.each do |c|
        unsatisfied.append(c) unless c.is_satisfied
      end

      constraints_consuming(v) do |c|
        c.recalculate
        todo.append(c.output)
      end
    end

    unsatisfied.sort { |c1, c2| c1.strength.stronger(c2.strength) }
    unsatisfied
  end

  def self.chain_test(n)
    # This is the standard DeltaBlue benchmark. A long chain of equality
    # constraints is constructed with a stay constraint on one end. An
    # edit constraint is then added to the opposite end and the time is
    # measured for adding and removing this constraint, and extracting
    # and executing a constraint satisfaction plan. There are two cases.
    # In case 1, the added constraint is stronger than the stay
    # constraint and values must propagate down the entire length of the
    # chain. In case 2, the added constraint is weaker than the stay
    # constraint so it cannot be accomodated. The cost in this case is,
    # of course, very low. Typical situations lie somewhere between these
    # two extremes.

    planner = Planner.new
    vars = Array.new(n + 1) { Variable.new }

    # thread a chain of equality constraints through the variables
    (0..(n - 1)).each do |i|
      v1 = vars[i]
      v2 = vars[i + 1]

      EqualityConstraint.new(v1, v2, SYM_REQUIRED, planner)
    end

    StayConstraint.new(vars.last, SYM_STRONG_DEFAULT, planner)
    edit = EditConstraint.new(vars.first, SYM_PREFERRED, planner)
    plan = planner.extract_plan_from_constraints([edit])

    (1..100).each do |v|
      vars.first.value = v
      plan.execute

      raise 'Chain test failed!!' if vars.last.value != v
    end

    edit.destroy_constraint(planner)
  end

  def self.projection_test(n)
    # This test constructs a two sets of variables related to each
    # other by a simple linear transformation (scale and offset). The
    # time is measured to change a variable on either side of the
    # mapping and to change the scale and offset factors.

    planner = Planner.new
    dests   = Vector.new
    scale   = Variable.value(10)
    offset  = Variable.value(1000)

    src = nil
    dst = nil

    (1..n).each do |i|
      src = Variable.value(i)
      dst = Variable.value(i)
      dests.append(dst)
      StayConstraint.new(src, SYM_DEFAULT, planner)
      ScaleConstraint.new(src, scale, offset, dst, SYM_REQUIRED, planner)
    end

    planner.change_var(src, 17)
    raise 'Projection 1 failed' if dst.value != 1170

    planner.change_var(dst, 1050)
    raise 'Projection 2 failed' if src.value != 5

    planner.change_var(scale, 5)
    (0..(n - 2)).each do |i|
      raise 'Projection 3 failed' if dests.at(i).value != ((i + 1) * 5 + 1000)
    end

    planner.change_var(offset, 2000)
    (0..(n - 2)).each do |i|
      raise 'Projection 4 failed' if dests.at(i).value != ((i + 1) * 5 + 2000)
    end
  end
end

class Sym
  attr_reader :custom_hash

  def initialize(hash)
    @custom_hash = hash
  end
end

SYM_ABSOLUTE_STRONGEST = Sym.new(0)
SYM_REQUIRED           = Sym.new(1)
SYM_STRONG_PREFERRED   = Sym.new(2)
SYM_PREFERRED          = Sym.new(3)
SYM_STRONG_DEFAULT     = Sym.new(4)
SYM_DEFAULT            = Sym.new(5)
SYM_WEAK_DEFAULT       = Sym.new(6)
SYM_ABSOLUTE_WEAKEST   = Sym.new(7)

class Strength
  attr_reader :arithmetic_value

  def initialize(strength_sym)
    @symbolic_value   = strength_sym
    @arithmetic_value = STRENGTH_TABLE.at(strength_sym)
  end

  def same_as(strength)
    @arithmetic_value == strength.arithmetic_value
  end

  def stronger(strength)
    @arithmetic_value < strength.arithmetic_value
  end

  def weaker(strength)
    @arithmetic_value > strength.arithmetic_value
  end

  def strongest(strength)
    if strength.stronger(self)
      strength
    else
      self
    end
  end

  def weakest(strength)
    if strength.weaker(self)
      strength
    else
      self
    end
  end

  def self.create_strength_table
    table = IdentityDictionary.new
    table.at_put(SYM_ABSOLUTE_STRONGEST, -10_000)
    table.at_put(SYM_REQUIRED,              -800)
    table.at_put(SYM_STRONG_PREFERRED,      -600)
    table.at_put(SYM_PREFERRED,             -400)
    table.at_put(SYM_STRONG_DEFAULT,        -200)
    table.at_put(SYM_DEFAULT,                  0)
    table.at_put(SYM_WEAK_DEFAULT,           500)
    table.at_put(SYM_ABSOLUTE_WEAKEST,    10_000)
    table
  end

  def self.create_strength_constants
    constants = IdentityDictionary.new
    STRENGTH_TABLE.keys.each do |strength_sym|
      constants.at_put(strength_sym, new(strength_sym))
    end
    constants
  end

  STRENGTH_TABLE     = create_strength_table
  STRENGTH_CONSTANTS = create_strength_constants

  def self.of(sym)
    STRENGTH_CONSTANTS.at(sym)
  end
end

class AbstractConstraint
  attr_reader :strength

  def initialize(strength_sym)
    @strength = Strength.of(strength_sym)
  end

  def is_input
    false
  end

  def add_constraint(planner)
    add_to_graph
    planner.incremental_add(self)
  end

  def destroy_constraint(planner)
    planner.incremental_remove(self) if is_satisfied
    remove_from_graph
  end

  def inputs_known(mark)
    !inputs_has_one { |v| !(v.mark == mark || v.stay || !v.determined_by) }
  end

  def satisfy(mark, planner)
    choose_method(mark)

    if is_satisfied
      inputs_do { |i| i.mark = mark }

      outx = output
      overridden = outx.determined_by

      overridden.mark_unsatisfied if overridden

      outx.determined_by = self

      unless planner.add_propagate(self, mark)
        raise 'Cycle encountered adding: Constraint removed'
      end

      outx.mark = mark
      overridden
    else
      if @strength.same_as(REQUIRED)
        raise 'Failed to satisfy a required constraint'
      end
      nil
    end
  end
end

class BinaryConstraint < AbstractConstraint
  def initialize(v1, v2, strength, _planner)
    super(strength)
    @v1 = v1
    @v2 = v2
    @direction = nil
  end

  def is_satisfied
    @direction != nil
  end

  def add_to_graph
    @v1.add_constraint(self)
    @v2.add_constraint(self)
    @direction = nil
  end

  def remove_from_graph
    @v1.remove_constraint(self) if @v1
    @v2.remove_constraint(self) if @v2

    @direction = nil
  end

  def choose_method(mark)
    if @v1.mark == mark
      if @v2.mark != mark && @strength.stronger(@v2.walk_strength)
        return @direction = :forward
      else
        return @direction = nil
      end
    end

    if @v2.mark == mark
      if @v1.mark != mark && @strength.stronger(@v1.walk_strength)
        return @direction = :backward
      else
        return @direction = nil
      end
    end

    if @v1.walk_strength.weaker(@v2.walk_strength)
      if @strength.stronger(@v1.walk_strength)
        @direction = :backward
      else
        @direction = nil
      end
    else
      if @strength.stronger(@v2.walk_strength)
        @direction = :forward
      else
        @direction = nil
      end
    end
  end

  def inputs_do
    if @direction == :forward
      yield @v1
    else
      yield @v2
    end
  end

  def inputs_has_one
    if @direction == :forward
      yield @v1
    else
      yield @v2
    end
  end

  def mark_unsatisfied
    @direction = nil
  end

  def output
    if @direction == :forward
      @v2
    else
      @v1
    end
  end

  def recalculate
    if @direction == :forward
      ihn  = @v1
      outx = @v2
    else
      ihn  = @v2
      outx = @v1
    end

    outx.walk_strength = @strength.weakest(ihn.walk_strength)
    outx.stay = ihn.stay

    execute if outx.stay
  end
end

class UnaryConstraint < AbstractConstraint
  attr_reader :output

  def initialize(v, strength, planner)
    super(strength)
    @output    = v
    @satisfied = false
    add_constraint(planner)
  end

  def is_satisfied
    @satisfied
  end

  def add_to_graph
    @output.add_constraint(self)
    @satisfied = false
  end

  def remove_from_graph
    @output.remove_constraint(self) if @output
    @satisfied = false
  end

  def choose_method(mark)
    @satisfied = @output.mark != mark &&
                 @strength.stronger(@output.walk_strength)
  end

  def inputs_do
    # No-op. I have no input variable.
  end

  def inputs_has_one
    false
  end

  def mark_unsatisfied
    @satisfied = false
  end

  def recalculate
    @output.walk_strength = @strength
    @output.stay          = !is_input

    execute if @output.stay
  end
end

class EditConstraint < UnaryConstraint
  def is_input
    true
  end

  def execute
    # Edit constraints does nothing.
  end
end

class EqualityConstraint < BinaryConstraint
  def initialize(var1, var2, strength, planner)
    super(var1, var2, strength, planner)
    add_constraint(planner)
  end

  def execute
    if @direction == :forward
      @v2.value = @v1.value
    else
      @v1.value = @v2.value
    end
  end
end

class ScaleConstraint < BinaryConstraint
  def initialize(src, scale, offset, dest, strength, planner)
    super(src, dest, strength, planner)
    @scale  = scale
    @offset = offset
    add_constraint(planner)
  end

  def add_to_graph
    @v1.add_constraint(self)
    @v2.add_constraint(self)
    @scale.add_constraint(self)
    @offset.add_constraint(self)
    @direction = nil
  end

  def remove_from_graph
    @v1.remove_constraint(self) if @v1
    @v2.remove_constraint(self) if @v2

    @scale.remove_constraint(self)  if @scale
    @offset.remove_constraint(self) if @offset

    @direction = nil
  end

  def execute
    if @direction == :forward
      @v2.value = @v1.value * @scale.value + @offset.value
    else
      @v1.value = (@v2.value - @offset.value) / @scale.value
    end
  end

  def inputs_do
    if @direction == :forward
      yield @v1
      yield @scale
      yield @offset
    else
      yield @v2
      yield @scale
      yield @offset
    end
  end

  def recalculate
    if @direction == :forward
      ihn  = @v1
      outx = @v2
    else
      outx = @v2
      ihn  = @v1
    end
    outx.walk_strength = @strength.weakest(ihn.walk_strength)
    outx.stay = (ihn.stay && @scale.stay && @offset.stay)

    execute if outx.stay
  end
end

class StayConstraint < UnaryConstraint
  def execute
    # Stay Constraints do nothing
  end
end

class Variable
  attr_accessor :value, :constraints, :determined_by, :walk_strength, :stay, :mark

  def initialize
    @value         = 0
    @constraints   = Vector.new(2)
    @determined_by = nil
    @walk_strength = ABSOLUTE_WEAKEST
    @stay          = true
    @mark          = 0
  end

  def add_constraint(constraint)
    @constraints.append(constraint)
  end

  def remove_constraint(constraint)
    @constraints.remove(constraint)

    @determined_by = nil if @determined_by == constraint
  end

  def self.value(initial_value)
    o = new
    o.value = initial_value
    o
  end
end

ABSOLUTE_STRONGEST = Strength.of(SYM_ABSOLUTE_STRONGEST)
ABSOLUTE_WEAKEST   = Strength.of(SYM_ABSOLUTE_WEAKEST)
REQUIRED           = Strength.of(SYM_REQUIRED)

File: benchmarks/JavaScript/deltablue.js
// @ts-check
// The benchmark in its current state is a derivation from the SOM version,
// which is derived from Mario Wolczko's Smalltalk version of DeltaBlue.
//
// The original license details are available here:
// http://web.archive.org/web/20050825101121/http://www.sunlabs.com/people/mario/java_benchmarking/index.html

const { Benchmark } = require('./benchmark');
const { Vector, IdentityDictionary } = require('./som');

class Plan extends Vector {
  constructor() {
    super(15);
  }

  execute() {
    this.forEach((c) => c.execute());
  }
}

class Sym {
  constructor(hash) {
    this.hash = hash;
  }

  customHash() {
    return this.hash;
  }
}

const ABSOLUTE_STRONGEST = new Sym(0);
const REQUIRED = new Sym(1);
const STRONG_PREFERRED = new Sym(2);
const PREFERRED = new Sym(3);
const STRONG_DEFAULT = new Sym(4);
const DEFAULT = new Sym(5);
const WEAK_DEFAULT = new Sym(6);
const ABSOLUTE_WEAKEST = new Sym(7);

function createStrengthTable() {
  const strengthTable = new IdentityDictionary();
  strengthTable.atPut(ABSOLUTE_STRONGEST, -10000);
  strengthTable.atPut(REQUIRED, -800);
  strengthTable.atPut(STRONG_PREFERRED, -600);
  strengthTable.atPut(PREFERRED, -400);
  strengthTable.atPut(STRONG_DEFAULT, -200);
  strengthTable.atPut(DEFAULT, 0);
  strengthTable.atPut(WEAK_DEFAULT, 500);
  strengthTable.atPut(ABSOLUTE_WEAKEST, 10000);
  return strengthTable;
}

class Strength {
  constructor(symbolicValue) {
    this.arithmeticValue = Strength.strengthTable.at(symbolicValue);
  }

  sameAs(s) {
    return this.arithmeticValue === s.arithmeticValue;
  }

  stronger(s) {
    return this.arithmeticValue < s.arithmeticValue;
  }

  weaker(s) {
    return this.arithmeticValue > s.arithmeticValue;
  }

  strongest(s) {
    return s.stronger(this) ? s : this;
  }

  weakest(s) {
    return s.weaker(this) ? s : this;
  }

  static of(strength) {
    return Strength.strengthConstant.at(strength);
  }

  static strengthTable = createStrengthTable();

  static createStrengthConstants() {
    const strengthConstant = new IdentityDictionary();
    Strength.strengthTable.getKeys().forEach((key) => {
      strengthConstant.atPut(key, new Strength(key));
    });
    return strengthConstant;
  }

  static strengthConstant = Strength.createStrengthConstants();

  static absoluteWeakest = Strength.of(ABSOLUTE_WEAKEST);

  static required = Strength.of(REQUIRED);
}

class AbstractConstraint {
  constructor(strengthSym) {
    this.strength = Strength.of(strengthSym);
  }

  isInput() {
    return false;
  }

  addConstraint(planner) {
    this.addToGraph();
    planner.incrementalAdd(this);
  }

  destroyConstraint(planner) {
    if (this.isSatisfied()) {
      planner.incrementalRemove(this);
    }
    this.removeFromGraph();
  }

  inputsKnown(mark) {
    return !this.inputsHasOne((v) => !(v.mark === mark || v.stay || v.determinedBy === null));
  }

  satisfy(mark, planner) {
    let overridden;
    this.chooseMethod(mark);

    if (this.isSatisfied()) {
      // constraint can be satisfied
      // mark inputs to allow cycle detection in addPropagate
      this.inputsDo((i) => { i.mark = mark; });

      const out = this.getOutput();
      overridden = out.determinedBy;
      if (overridden !== null) {
        overridden.markUnsatisfied();
      }
      out.determinedBy = this;
      if (!planner.addPropagate(this, mark)) {
        throw new Error('Cycle encountered');
      }
      out.mark = mark;
    } else {
      overridden = null;
      if (this.strength.sameAs(Strength.required)) {
        throw new Error('Could not satisfy a required constraint');
      }
    }
    return overridden;
  }
}

class BinaryConstraint extends AbstractConstraint {
  // eslint-disable-next-line no-unused-vars
  constructor(var1, var2, strength, planner) {
    super(strength);
    this.v1 = var1;
    this.v2 = var2;
    this.direction = null;
  }

  isSatisfied() {
    return this.direction !== null;
  }

  addToGraph() {
    this.v1.addConstraint(this);
    this.v2.addConstraint(this);
    this.direction = null;
  }

  removeFromGraph() {
    if (this.v1 !== null) {
      this.v1.removeConstraint(this);
    }
    if (this.v2 !== null) {
      this.v2.removeConstraint(this);
    }
    this.direction = null;
  }

  chooseMethod(mark) {
    if (this.v1.mark === mark) {
      if (this.v2.mark !== mark && this.strength.stronger(this.v2.walkStrength)) {
        this.direction = 'forward';
        return this.direction;
      }
      this.direction = null;
      return this.direction;
    }

    if (this.v2.mark === mark) {
      if (this.v1.mark !== mark && this.strength.stronger(this.v1.walkStrength)) {
        this.direction = 'backward';
        return this.direction;
      }
      this.direction = null;
      return this.direction;
    }

    // If we get here, neither variable is marked, so we have a choice.
    if (this.v1.walkStrength.weaker(this.v2.walkStrength)) {
      if (this.strength.stronger(this.v1.walkStrength)) {
        this.direction = 'backward';
        return this.direction;
      }
      this.direction = null;
      return this.direction;
    }
    if (this.strength.stronger(this.v2.walkStrength)) {
      this.direction = 'forward';
      return this.direction;
    }
    this.direction = null;
    return this.direction;
  }

  inputsDo(fn) {
    if (this.direction === 'forward') {
      fn(this.v1);
    } else {
      fn(this.v2);
    }
  }

  inputsHasOne(fn) {
    if (this.direction === 'forward') {
      return fn(this.v1);
    }
    return fn(this.v2);
  }

  markUnsatisfied() {
    this.direction = null;
  }

  getOutput() {
    return this.direction === 'forward' ? this.v2 : this.v1;
  }

  recalculate() {
    let ihn;
    let out;

    if (this.direction === 'forward') {
      ihn = this.v1; out = this.v2;
    } else {
      ihn = this.v2; out = this.v1;
    }

    out.walkStrength = this.strength.weakest(ihn.walkStrength);
    out.stay = ihn.stay;
    if (out.stay) {
      this.execute();
    }
  }
}

class UnaryConstraint extends AbstractConstraint {
  constructor(v, strength, planner) {
    super(strength);
    this.output = v;
    this.satisfied = false;

    this.addConstraint(planner);
  }

  isSatisfied() {
    return this.satisfied;
  }

  addToGraph() {
    this.output.addConstraint(this);
    this.satisfied = false;
  }

  removeFromGraph() {
    if (this.output !== null) {
      this.output.removeConstraint(this);
    }
    this.satisfied = false;
  }

  chooseMethod(mark) {
    this.satisfied = this.output.mark !== mark
      && this.strength.stronger(this.output.walkStrength);
    return null;
  }

  // eslint-disable-next-line no-unused-vars
  inputsDo(fn) {
    // I have no input variables
  }

  // eslint-disable-next-line no-unused-vars
  inputsHasOne(fn) {
    return false;
  }

  markUnsatisfied() {
    this.satisfied = false;
  }

  getOutput() {
    return this.output;
  }

  recalculate() {
    this.output.walkStrength = this.strength;
    this.output.stay = !this.isInput();
    if (this.output.stay) {
      this.execute(); // stay optimization
    }
  }
}

class EditConstraint extends UnaryConstraint {
  isInput() {
    return true;
  }

  execute() {}
}

class EqualityConstraint extends BinaryConstraint {
  constructor(var1, var2, strength, planner) {
    super(var1, var2, strength, planner);
    this.addConstraint(planner);
  }

  execute() {
    if (this.direction === 'forward') {
      this.v2.value = this.v1.value;
    } else {
      this.v1.value = this.v2.value;
    }
  }
}

class ScaleConstraint extends BinaryConstraint {
  constructor(src, scale, offset, dest, strength, planner) {
    super(src, dest, strength, planner);
    this.scale = scale;
    this.offset = offset;

    this.addConstraint(planner);
  }

  addToGraph() {
    this.v1.addConstraint(this);
    this.v2.addConstraint(this);
    this.scale.addConstraint(this);
    this.offset.addConstraint(this);
    this.direction = null;
  }

  removeFromGraph() {
    if (this.v1 !== null) { this.v1.removeConstraint(this); }
    if (this.v2 !== null) { this.v2.removeConstraint(this); }
    if (this.scale !== null) { this.scale.removeConstraint(this); }
    if (this.offset !== null) { this.offset.removeConstraint(this); }
    this.direction = null;
  }

  execute() {
    if (this.direction === 'forward') {
      this.v2.value = this.v1.value * this.scale.value + this.offset.value;
    } else {
      this.v1.value = (this.v2.value - this.offset.value) / this.scale.value;
    }
  }

  inputsDo(fn) {
    if (this.direction === 'forward') {
      fn(this.v1);
      fn(this.scale);
      fn(this.offset);
    } else {
      fn(this.v2);
      fn(this.scale);
      fn(this.offset);
    }
  }

  recalculate() {
    let ihn;
    let out;

    if (this.direction === 'forward') {
      ihn = this.v1; out = this.v2;
    } else {
      out = this.v1; ihn = this.v2;
    }

    out.walkStrength = this.strength.weakest(ihn.walkStrength);
    out.stay = ihn.stay && this.scale.stay && this.offset.stay;
    if (out.stay) {
      this.execute(); // stay optimization
    }
  }
}

class StayConstraint extends UnaryConstraint {
  execute() {}
}

class Variable {
  constructor() {
    this.value = 0;
    this.constraints = new Vector(2);
    this.determinedBy = null;
    this.walkStrength = Strength.absoluteWeakest;
    this.stay = true;
    this.mark = 0;
  }

  addConstraint(c) {
    this.constraints.append(c);
  }

  removeConstraint(c) {
    this.constraints.remove(c);
    if (this.determinedBy === c) {
      this.determinedBy = null;
    }
  }

  static value(aValue) {
    const v = new Variable();
    v.value = aValue;
    return v;
  }
}

class Planner {
  constructor() {
    this.currentMark = 1;
  }

  newMark() {
    this.currentMark += 1;
    return this.currentMark;
  }

  incrementalAdd(c) {
    const mark = this.newMark();
    let overridden = c.satisfy(mark, this);

    while (overridden !== null) {
      overridden = overridden.satisfy(mark, this);
    }
  }

  incrementalRemove(c) {
    const out = c.getOutput();
    c.markUnsatisfied();
    c.removeFromGraph();

    const unsatisfied = this.removePropagateFrom(out);
    unsatisfied.forEach((u) => this.incrementalAdd(u));
  }

  extractPlanFromConstraints(constraints) {
    const sources = new Vector();

    constraints.forEach((c) => {
      if (c.isInput() && c.isSatisfied()) {
        sources.append(c);
      }
    });

    return this.makePlan(sources);
  }

  makePlan(sources) {
    const mark = this.newMark();
    const plan = new Plan();
    const todo = sources;

    while (!todo.isEmpty()) {
      const c = todo.removeFirst();

      if (c.getOutput().mark !== mark && c.inputsKnown(mark)) {
        // not in plan already and eligible for inclusion
        plan.append(c);
        c.getOutput().mark = mark;
        this.addConstraintsConsumingTo(c.getOutput(), todo);
      }
    }
    return plan;
  }

  propagateFrom(v) {
    const todo = new Vector();
    this.addConstraintsConsumingTo(v, todo);

    while (!todo.isEmpty()) {
      const c = todo.removeFirst();
      c.execute();
      this.addConstraintsConsumingTo(c.getOutput(), todo);
    }
  }

  addConstraintsConsumingTo(v, coll) {
    const determiningC = v.determinedBy;

    v.constraints.forEach((c) => {
      if (c !== determiningC && c.isSatisfied()) {
        coll.append(c);
      }
    });
  }

  addPropagate(c, mark) {
    const todo = Vector.with(c);

    while (!todo.isEmpty()) {
      const d = todo.removeFirst();

      if (d.getOutput().mark === mark) {
        this.incrementalRemove(c);
        return false;
      }
      d.recalculate();
      this.addConstraintsConsumingTo(d.getOutput(), todo);
    }
    return true;
  }

  change(v, newValue) {
    const editC = new EditConstraint(v, PREFERRED, this);
    const editV = Vector.with(editC);
    const plan = this.extractPlanFromConstraints(editV);

    for (let i = 0; i < 10; i += 1) {
      v.value = newValue;
      plan.execute();
    }
    editC.destroyConstraint(this);
  }

  constraintsConsuming(v, fn) {
    const determiningC = v.determinedBy;
    v.constraints.forEach((c) => {
      if (c !== determiningC && c.isSatisfied()) {
        fn(c);
      }
    });
  }

  removePropagateFrom(out) {
    const unsatisfied = new Vector();

    out.determinedBy = null;
    out.walkStrength = Strength.absoluteWeakest;
    out.stay = true;

    const todo = Vector.with(out);

    while (!todo.isEmpty()) {
      const v = todo.removeFirst();

      v.constraints.forEach((c) => {
        if (!c.isSatisfied()) { unsatisfied.append(c); }
      });

      this.constraintsConsuming(v, (c) => {
        c.recalculate();
        todo.append(c.getOutput());
      });
    }

    unsatisfied.sort((c1, c2) => c1.strength.stronger(c2.strength));
    return unsatisfied;
  }

  static chainTest(n) {
    const planner = new Planner();
    const vars = new Array(n + 1);

    for (let i = 0; i < n + 1; i += 1) {
      vars[i] = new Variable();
    }

    // Build chain of n equality constraints
    for (let i = 0; i < n; i += 1) {
      const v1 = vars[i];
      const v2 = vars[i + 1];
      new EqualityConstraint(v1, v2, REQUIRED, planner);
    }

    new StayConstraint(vars[n], STRONG_DEFAULT, planner);

    const editC = new EditConstraint(vars[0], PREFERRED, planner);
    const editV = Vector.with(editC);
    const plan = planner.extractPlanFromConstraints(editV);

    for (let i = 0; i < 100; i += 1) {
      vars[0].value = i;
      plan.execute();
      if (vars[n].value !== i) {
        throw new Error('Chain test failed!');
      }
    }
    editC.destroyConstraint(planner);
  }

  static projectionTest(n) {
    const planner = new Planner();
    const dests = new Vector();
    const scale = Variable.value(10);
    const offset = Variable.value(1000);

    let src = null;
    let dst = null;

    for (let i = 1; i <= n; i += 1) {
      src = Variable.value(i);
      dst = Variable.value(i);
      dests.append(dst);
      new StayConstraint(src, DEFAULT, planner);
      new ScaleConstraint(src, scale, offset, dst, REQUIRED, planner);
    }

    planner.change(src, 17);
    if (dst.value !== 1170) {
      throw new Error('Projection test 1 failed!');
    }

    planner.change(dst, 1050);
    if (src.value !== 5) {
      throw new Error('Projection test 2 failed!');
    }

    planner.change(scale, 5);
    for (let i = 0; i < n - 1; i += 1) {
      if (dests.at(i).value !== (i + 1) * 5 + 1000) {
        throw new Error('Projection test 3 failed!');
      }
    }

    planner.change(offset, 2000);
    for (let i = 0; i < n - 1; i += 1) {
      if (dests.at(i).value !== (i + 1) * 5 + 2000) {
        throw new Error('Projection test 4 failed!');
      }
    }
  }
}

class DeltaBlue extends Benchmark {
  innerBenchmarkLoop(innerIterations) {
    Planner.chainTest(innerIterations);
    Planner.projectionTest(innerIterations);
    return true;
  }
}

exports.newInstance = () => new DeltaBlue();
